compiler/rustc_lint/src/builtin.rs

```diff
@@ -1699,32 +1699,57 @@ impl EarlyLintPass for EllipsisInclusiveRangePatterns {
             let suggestion = "use `..=` for an inclusive range";
             if parenthesise {
                 self.node_id = Some(pat.id);
-                cx.struct_span_lint(ELLIPSIS_INCLUSIVE_RANGE_PATTERNS, pat.span, |lint| {
-                    let end = expr_to_string(&end);
-                    let replace = match start {
-                        Some(start) => format!("&({}..={})", expr_to_string(&start), end),
-                        None => format!("&(..={})", end),
-                    };
-                    lint.build(msg)
-                        .span_suggestion(
-                            pat.span,
-                            suggestion,
-                            replace,
-                            Applicability::MachineApplicable,
-                        )
-                        .emit();
-                });
+                let end = expr_to_string(&end);
+                let replace = match start {
+                    Some(start) => format!("&({}..={})", expr_to_string(&start), end),
+                    None => format!("&(..={})", end),
+                };
+                if cx.sess().edition() >= Edition::Edition2021 {
+                    let mut err =
+                        rustc_errors::struct_span_err!(cx.sess, pat.span, E0783, "{}", msg,);
+                    err.span_suggestion(
+                        pat.span,
+                        suggestion,
+                        replace,
+                        Applicability::MachineApplicable,
+                    )
+                    .emit();
+                } else {
+                    cx.struct_span_lint(ELLIPSIS_INCLUSIVE_RANGE_PATTERNS, pat.span, |lint| {
+                        lint.build(msg)
+                            .span_suggestion(
+                                pat.span,
+                                suggestion,
+                                replace,
+                                Applicability::MachineApplicable,
+                            )
+                            .emit();
+                    });
+                }
             } else {
-                cx.struct_span_lint(ELLIPSIS_INCLUSIVE_RANGE_PATTERNS, join, |lint| {
-                    lint.build(msg)
-                        .span_suggestion_short(
-                            join,
-                            suggestion,
-                            "..=".to_owned(),
-                            Applicability::MachineApplicable,
-                        )
-                        .emit();
-                });
+                let replace = "..=".to_owned();
+                if cx.sess().edition() >= Edition::Edition2021 {
+                    let mut err =
+                        rustc_errors::struct_span_err!(cx.sess, pat.span, E0783, "{}", msg,);
```

Comment:
Note: lint passes _typically_ do not have side effects, if a lint is disabled there should ideally be no behavior differences between running the pass or not (and that's a potential optimization to do in the future).

I'm not sure how strictly the compiler team follows this; it's possible that when we do such optimizations we'll have to go back and audit our lints (which is fine!), but I just want to flag this.
I'm _not sure_ how str

Reply:
The only concern is that performing this check in the appropriate place as if we had no editions could cause divergence in the behavior. Emitting the error in the same place as we currently emit the lint _should_ make the transition less painful than if we start denying in places where we currently don't even lint. That being said, this _could_ be accepted as acceptable breakage to bestow on the ecosystem in the name of codebase cleanliness.

Reply:
Hmm, I missed this discussion the first time around. A few thoughts:

* Maybe we want some way to have the lint pass indicate that it is *required*? I'm imagining adding a `fn required(&self) -> bool { false }` to the trait, and this lint could return true. This might be convenient if we commonly wind up with things that are errors in one edition and lints in another.
* Alternatively, we could maybe add a group of lints, not 'late lints' but 'required late lints' and put the lint into that group, something like that.

Moving the logic out from the "lint helper" just seems like a bit of extra work, I'm not sure what would be a nice place to put it, we'd have to reproduce the "traverse the tree" code. That said, I don't object to it.

Reply:
I'm also not quite following. That said, I don't think we hold ourselves to a sufficiently high standard here *today* that I would be comfortable applying a "no pass" optimization if lints were disabled. We also merge all built-in lint passes into a few groups (by when they need to run), so there's not really an actual "per-lint" full traversal of any tree, so it's not obvious a major win is to be had. (But I don't think we've profiled much).

I think this seems OK for now.

Reply:
As long as we have tests for the hard error, I'm comfortable with it.